node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+			if let Some(hash) = activated.get(0) {
+				state.live_block_hash = *hash;
+			}
+
+			Ok(false)
+		}
+		_ => Ok(false),
+	}
+}
+
+/// Report a reputation change for a peer.
+async fn report_peer(
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	peer: PeerId,
+	rep: Rep,
+) {
+	ctx.send_message(AllMessages::NetworkBridge(NetworkBridgeMessage::ReportPeer(peer, rep))).await;
+}
+
+/// Machinery around launching interactions into the background.
+#[tracing::instrument(level = "trace", skip(ctx, state), fields(subsystem = LOG_TARGET))]
+async fn launch_interaction(
+	state: &mut State,
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	session_index: SessionIndex,
+	session_info: SessionInfo,
+	receipt: CandidateReceipt,
+	response_sender: oneshot::Sender<Result<AvailableData, RecoveryError>>,
+) -> error::Result<()> {
+	use rand::seq::SliceRandom;
+	use rand::thread_rng;
+
+	let threshold = recovery_threshold(session_info.validators.len());
+	let to_state = state.from_interaction_tx.clone();
+	let candidate_hash = receipt.hash();
+	let erasure_root = receipt.descriptor.erasure_root;
+	let validators = session_info.validators.clone();
+	let mut shuffling: Vec<_> = (0..validators.len() as u32).collect();
+
+	state.interactions.insert(
+		candidate_hash.clone(),
+		InteractionHandle {
+			awaiting: vec![response_sender],
+		}
+	);
+
+	{
+		// make borrow checker happy.
+		let mut rng = thread_rng();
+		shuffling.shuffle(&mut rng);
+	}
+
+	let interaction = Interaction {
+		to_state,
+		validators,
+		shuffling,
+		threshold,
+		candidate_hash,
+		erasure_root,
+		received_chunks: HashMap::new(),
+		requesting_chunks: FuturesUnordered::new(),
+	};
+
+	if let Err(e) = ctx.spawn("recovery interaction", interaction.run().boxed()).await {
+		tracing::warn!(
+			target: LOG_TARGET,
+			err = ?e,
+			"Failed to spawn a recovery interaction task",
+		);
+	}
+
+	Ok(())
+}
+
+/// Handles an availability recovery request.
+#[tracing::instrument(level = "trace", skip(ctx, state), fields(subsystem = LOG_TARGET))]
+async fn handle_recover(
+	state: &mut State,
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	receipt: CandidateReceipt,
+	response_sender: oneshot::Sender<Result<AvailableData, RecoveryError>>,
+) -> error::Result<()> {
+	let candidate_hash = receipt.hash();
+
+	if let Some(result) = state.availability_lru.get(&candidate_hash) {
+		if let Err(e) = response_sender.send(result.clone()) {
+			tracing::warn!(
+				target: LOG_TARGET,
+				err = ?e,
+				"Error responding with an availability recovery result",
+			);
+		}
+		return Ok(());
+	}
+
+	if let Some(interaction) = state.interactions.get_mut(&candidate_hash) {
+		interaction.awaiting.push(response_sender);
+		return Ok(());
+	}
+
+	let session_index = request_session_index_for_child_ctx(state.live_block_hash, ctx).await?.await??;
```

Comment:
we're supposed to get the session index of the candidate, not the session index of the most recent block. might not actually be that easily available here, but the caller of the subsystem should have it. can alter API to accept a `SessionIndex` as a parameter.